File: src/main/java/fastandslowpointers/circulararray/CircularArrayLoop.java
package fastandslowpointers.circulararray;

public class CircularArrayLoop {

    public boolean isCircularArrayLoop(int[] nums) {

        for (int i = 0; i < nums.length; i++) {
            boolean isPositive = nums[i] >= 0;
            int slow = i;
            int fast = i;
            while (true) {
                slow = isPositive ? (slow + nums[slow]) % nums.length : (slow + nums[slow] + nums.length) % nums.length;
                if (nums[slow] >= 0 ^ isPositive) break;
                fast = isPositive ? (fast + nums[fast]) % nums.length : (fast + nums[fast] + nums.length) % nums.length;
                if (nums[fast] >= 0 ^ isPositive) break;
                fast = isPositive ? (fast + nums[fast]) % nums.length : (fast + nums[fast] + nums.length) % nums.length;
                if (nums[fast] >= 0 ^ isPositive) break;
                if (fast == slow) return true;
            }
        }
        return false;
    }

}

File: src/main/kotlin/fastandslowpointers/circulararray/CircularArrayLoop.kt
package fastandslowpointers.circulararray

fun circularArrayLoop(nums: IntArray): Boolean {

    nums.forEachIndexed { i, n ->
        val isPositive = n >= 0
        var slow = i
        var fast = i
        while (true) {
            slow = if (isPositive) (slow + nums[slow]) % nums.size else (slow + nums[slow] + nums.size) % nums.size
            if ((nums[slow] >= 0) xor isPositive) break
            fast = if (isPositive) (fast + nums[fast]) % nums.size else (fast + nums[fast] + nums.size) % nums.size
            if ((nums[fast] >= 0) xor isPositive) break
            fast = if (isPositive) (fast + nums[fast]) % nums.size else (fast + nums[fast] + nums.size) % nums.size
            if ((nums[fast] >= 0) xor isPositive) break
            if (fast == slow) return true
        }
    }
    return false

}
